storage/remote/queue_manager.go

```diff
@@ -302,11 +341,83 @@ func NewQueueManager(reg prometheus.Registerer, metrics *queueManagerMetrics, lo
 	}
 
 	t.watcher = wal.NewWatcher(reg, wal.NewWatcherMetrics(reg), logger, client.Name(), t, walDir)
+	if t.cfg.SendMetadata {
+		t.metadataWatcher = scrape.NewMetadataWatcher(logger, sm, client.Name(), t, t.cfg.MetadataSendInterval)
+	}
 	t.shards = t.newShards()
 
 	return t
 }
 
+// AppendMetadata queues metadata to be sent to the remote storage.
+func (t *QueueManager) AppendMetadata(metadataList []scrape.MetricMetadata) bool {
+	metadataSet := map[scrape.MetricMetadata]struct{}{}
+
+	for _, metric := range metadataList {
+		metadataSet[metric] = struct{}{}
+	}
+
+	metadata := []*prompb.MetricMetadata{}
```

Comment:
Can you make this not a `[]*type` but instead a `[]type`?  There are options in gogoproto for this - it saves on allocations massively.

Also, can you `make([]type, 0, len(metadataList)` - again, saving on allocations.